### PR TITLE
fix(pi07_paligemma): backport prefix gating + cross-att length to feat/pi07 (#226)

### PR DESCRIPTION
**Pull request by shuheng-liu**
## What this does

Backports the combined [#205](https://github.com/TensorAuto/OpenTau/pull/205) + [#226](https://github.com/TensorAuto/OpenTau/issues/226) gating fixes from `fix/pi07_paligemma_fixes` into `feat/pi07`'s pi07_paligemma low-level planner. The full set of bugs that #205 + #226 address are present on `feat/pi07` today (only a partial earlier attempt landed via `2949e73 "Applying pi07_paligemma fixes to pi07"`).

`PI07LowLevelPlannerFlowMatching.embed_prefix` (in `src/opentau/policies/pi07_paligemma/low_level_planner/modeling_pi07_low_level.py`):

- **Content gating (#205).** Wrap the response, subgoal (header + image loop + footer), and metadata blocks in `if has_response / has_subgoal / has_metadata` guards computed at the top of `embed_prefix` from the actual masks. Without this, with `*_drop_prob=1.0` every prefix carried real (non-padded) spurious tokens with `att_masks=[1,...]` causal-block boundaries that corrupted cumsum for every following token. Subgoal header / footer pad-masks are now per-sample via `sample_has_subgoal` so mixed batches stay correct.
- **Separator gating (#226).** State-end now switches between `", "` (when an optional follows) and `":\n"` (when nothing does) — matching pi05's all-dropped layout. Prefix-end (`";\n "`) is wrapped in `if has_any_optional:` so it's skipped entirely when there's no optional list to terminate.
- **`"Action: "` indicator att_masks (#226).** Changed from `[1] + [0]*(N-1)` (one bidirectional block) to `[1]*N` (one causal block per token). The embeddings already matched pi05 byte-for-byte, but `make_att_2d_masks` shifted every following discrete-action position's cumsum by `N-1`, silently moving the `prefix_out` boundary the discrete-action CE reads.
- **`forward()` cross-attention length (#205).** `num_cross_att_tokens` and `prefix_offsets` now subtract both the `"Action: "` indicator length and `discrete_action_max_length`, matching pi05's `discrete_action_indicator_max_length` logic so the action expert sees the same prefix length at train and at inference (where neither block is in the prefix).
- **`prepare_subgoal_images` NameError (#205).** Track `last_subgoal_img` / `last_mask` inside the present-keys loop and guard the `empty_cameras` fallback with `if last_subgoal_img is not None and last_mask is not None:`. Default `subgoal_is_pad` initializer flips from `torch.zeros` to `torch.ones` (a missing key means "padded", not "real").
- **`prepare_metadata` defaults.** Same `torch.zeros` → `torch.ones` flip for `speed_is_pad` / `quality_is_pad` / `mistake_is_pad`.

Also cherry-picked from main: [#209](https://github.com/TensorAuto/OpenTau/pull/209) (CPU CI `--ignore` for the GPU+slow integration test) and [#217](https://github.com/TensorAuto/OpenTau/pull/217) (GPU CI `--ignore` + comment dedupe). These workflow changes are already on `main` but had not propagated to `feat/pi07`, leaving CPU CI broken at test collection because of the unrelated pre-existing `VJEPA2VideoEncoder` import (TODO #210 / blocked on #192).

**Out of scope** (separate concerns on `fix/pi07_paligemma_fixes` worth handling in their own PRs against `feat/pi07`):
- V-JEPA2 → SpaceTime SigLIP video-encoder swap.
- The `";\n "` → `":\n"` prefix-end string cleanup (commit `6e3f82a`).
- Any change to the gemma3 pi07 (the *other* policy on this branch) or the high-level planner.

## How it was tested

- `pre-commit run --files src/.../modeling_pi07_low_level.py tests/.../test_pi07_paligemma_low_level_planner.py` — all hooks pass (ruff, ruff-format, pyupgrade, gitleaks, bandit, etc.).
- `PYTHONPATH=src pytest tests/policies/ -m "not gpu" --ignore=tests/policies/test_pi07_paligemma_low_level_planner.py -q` → **125 passed, 2 skipped, 19 deselected** in 152s. (PYTHONPATH override needed locally because the worktree's pi07/ module isn't installed in the parent venv; CI's `uv sync` step makes this unnecessary.)
- The integration test `tests/policies/test_pi07_paligemma_low_level_planner.py::test_complete_pi07_low_level_pipeline` is `@pytest.mark.gpu @pytest.mark.slow` and ignored in both CPU and GPU CI per #209/#217 (TODO #210 / blocked on #192). The change does **not** alter prefix shapes for the all-populated path that test exercises; it only changes the `"Action: "` indicator's `att_masks` pattern, which is captured and re-fed to `make_att_2d_masks`, so `_verify_vlm_attention_mask` continues to hold. `_verify_position_ids` and `_verify_action_expert_attention_mask` were updated to reflect the new `forward()` formulas.
- **Not run:** GPU integration test (no GPU on this dev machine and the test is CI-ignored anyway). A reviewer with GPU access running the test manually would also need to drop the `--ignore` for that one file.

## How to checkout & try? (for the reviewer)

```bash
gh pr checkout 228
uv sync --extra dev --extra libero
pre-commit run --all-files
pytest -m "not gpu" -n auto
```

To exercise the integration test on a GPU box (oracle / mlbox):

```bash
pytest -m "gpu" -sx tests/policies/test_pi07_paligemma_low_level_planner.py::TestPI07LowLevelPlannerIntegration::test_complete_pi07_low_level_pipeline
```

To sanity-check the gating fires (Python REPL on a box where `pi05_mem.video_encoder.VJEPA2VideoEncoder` exists, or once #192 lands):

```python
# Build a minimal batch with no response/subgoal/metadata keys, call
# model.embed_prefix(...) directly, and confirm:
#   - prefix length shrinks by (response_max_length + subgoal_lead +
#     n_subgoal_cams * subgoal_tokens + comma + metadata_max_length +
#     prefix_end) tokens vs the all-populated case
#   - state-end is ":\n" instead of ", "
```

## Checklist

- [x] I have added Google-style docstrings to important functions and ensured function parameters are typed.
- [x] My PR includes policy-related changes.
  - [ ] If the above is checked: I have run the GPU pytests (pytest -m "gpu") and regression tests.

> GPU pytests checkbox left unchecked because the only relevant GPU test (`test_complete_pi07_low_level_pipeline`) is CI-ignored at import time on `feat/pi07` (pre-existing `VJEPA2VideoEncoder` issue, TODO #210). Happy to run it manually if a reviewer wants — please flag.

### Note: Before submitting this PR, please read the [contributor guideline](https://github.com/TensorAuto/OpenTau/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by shuheng-liu** [comment]
Wrong PR. My intention was to fix the issue in `p07` policy but this touches `pi07_paligemma` instead. Will open a new one
